ecosystem/typescript/sdk/src/tests/unit/builder_utils.test.ts

```diff
@@ -268,33 +292,131 @@ describe("BuilderUtils", () => {
+    expect(serializer.getBytes()).toEqual(new Uint8Array([0x3, 0x61, 0x62, 0x63]));
+  });
+
+  it("serializes an empty option arg", async () => {
+    let serializer = new Serializer();
+    serializeArg(undefined, new TypeTagStruct(optionStructTag(new TypeTagU8())), serializer);
+    expect(serializer.getBytes()).toEqual(new Uint8Array([0x0]));
+  });
+
+  it("serializes an option num arg", async () => {
+    let serializer = new Serializer();
+    serializeArg("1", new TypeTagStruct(optionStructTag(new TypeTagU8())), serializer);
+    expect(serializer.getBytes()).toEqual(new Uint8Array([0x1, 0x1]));
+  });
+
+  it("serializes an option string arg", async () => {
+    let serializer = new Serializer();
+    serializeArg("abc", new TypeTagStruct(optionStructTag(new TypeTagStruct(stringStructTag))), serializer);
+    expect(serializer.getBytes()).toEqual(new Uint8Array([0x1, 0x3, 0x61, 0x62, 0x63]));
+  });
+
+  it("serializes a optional Object", async () => {
     let serializer = new Serializer();
     serializeArg(
-      "abc",
-      new TypeTagStruct(
-        new StructTag(AccountAddress.fromHex("0x1"), new Identifier("string"), new Identifier("String"), []),
-      ),
+      "0x01",
+      new TypeTagStruct(optionStructTag(new TypeTagStruct(objectStructTag(new TypeTagU8())))),
       serializer,
     );
-    expect(serializer.getBytes()).toEqual(new Uint8Array([0x3, 0x61, 0x62, 0x63]));
+    //00 00 00 00 00000000 00000000 00000000 00000000 00000000 00000000 00000000
+    expect(serializer.getBytes()).toEqual(
+      new Uint8Array([
+        0x1, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+      ]),
+    );
+  });
 
-    serializer = new Serializer();
+  it("throws when nested type too high", async () => {
+    let serializer = new Serializer();
+    expect(() => {
+      serializeArg(
+        "abc",
+        new TypeTagStruct(
+          optionStructTag(
+            new TypeTagStruct(
+              optionStructTag(
+                new TypeTagStruct(
+                  optionStructTag(
+                    new TypeTagStruct(
+                      optionStructTag(
+                        new TypeTagStruct(
+                          optionStructTag(
+                            new TypeTagStruct(
+                              optionStructTag(
+                                new TypeTagStruct(
+                                  optionStructTag(
+                                    new TypeTagStruct(
+                                      optionStructTag(
+                                        new TypeTagStruct(
+                                          optionStructTag(new TypeTagStruct(objectStructTag(new TypeTagU8()))),
+                                        ),
+                                      ),
+                                    ),
```

Comment:
some people like to call recursive functions or other things that don't make this so ugly... can't we have some fun with lambdas / closures? why do you hate aesthetics ....

Reply:
I don't really know typescript well, but we can improve it